lib/surrealist/builder.rb

```diff
@@ -3,33 +3,87 @@
-          else
-            ValueAssigner.assign(schema: Schema.new(schema_key, schema_value),
-                                 instance: instance) { |coerced_value| schema[schema_key] = coerced_value }
-          end
+    attr_reader :carrier, :instance, :schema
+
+    # @param [Carrier] carrier instance of Surrealist::Carrier
+    # @param [Hash] schema the schema defined in the object's class.
+    # @param [Object] instance the instance of the object which methods from the schema are called on.
+    def initialize(carrier:, schema:, instance:)
+      @carrier = carrier
+      @schema = schema
+      @instance = instance
+    end
+
+    # A method that goes recursively through the schema hash, defines the values and type-checks them.
+    #
+    # @param [Hash] schema the schema defined in the object's class.
+    # @param [Object] instance the instance of the object which methods from the schema are called on.
+    #
+    # @raise +Surrealist::UndefinedMethodError+ if a key defined in the schema
+    #   does not have a corresponding method on the object.
+    #
+    # @return [Hash] a hash that will be dumped into JSON.
+    def call(schema: @schema, instance: @instance)
+      schema.each do |schema_key, schema_value|
+        if schema_value.is_a?(Hash)
+          check_for_ar(schema, instance, schema_key, schema_value)
+        else
+          ValueAssigner.assign(schema:   Schema.new(schema_key, schema_value),
+                               instance: instance) { |coerced_value| schema[schema_key] = coerced_value }
         end
-      rescue NoMethodError => e
-        raise Surrealist::UndefinedMethodError,
-              "#{e.message}. You have probably defined a key " \
-              "in the schema that doesn't have a corresponding method."
       end
+    rescue NoMethodError => e
+      Surrealist::ExceptionRaiser.raise_invalid_key!(e)
+    end
+
+    private
 
-      Schema = Struct.new(:key, :value)
+    # Checks if result is an instance of ActiveRecord::Relation
+    #
+    # @param [Hash] schema the schema defined in the object's class.
+    # @param [Object] instance the instance of the object which methods from the schema are called on.
+    # @param [Symbol] key the symbol that represents method on the instance
+    # @param [Any] value returned when key is called on instance
+    #
+    # @return [Hash] the schema hash
+    def check_for_ar(schema, instance, key, value)
+      if ar_collection?(instance, key)
+        construct_collection(schema, instance, key, value)
+      else
+        call(schema:   value,
+             instance: instance.respond_to?(key) ? instance.send(key) : instance)
+      end
+    end
+
+    # Checks if the instance responds to the method and whether it returns an AR::Relation
+    #
+    # @param [Object] instance
+    # @param [Symbol] method
+    #
+    # @return [Boolean]
+    def ar_collection?(instance, method)
```

Comment:
I tried to remove this...but with no success, this will probably become pretty nasty with more integrations. Seems to be caused only when joining(?)

Reply:
Yeah, that's unfortunate, but I didn't find a solution to escape this check. It's caused by `ActiveRecord::Associations::AssociationProxy`